src/hashupdater.cpp

```diff
@@ -5,19 +5,47 @@
  */
 
 #include "hashupdater.hpp"
-#include "util.hpp"
+#include "util/threadpool.hpp"
 #include "world.hpp"
 
 PasswordHashUpdater::PasswordHashUpdater(Config& config, const std::unordered_map<HashFunc, std::shared_ptr<Hasher>>& passwordHashers)
     : _config(config)
     , _passwordHashers(passwordHashers)
-    , _database(nullptr)
-    , _terminating(false)
-    , _updateSem(0)
 {
-    this->_updateThread = std::thread([this] { this->updateThreadProc(); });
+}
+
+void PasswordHashUpdater::QueueUpdatePassword(const std::string& username, util::secure_string&& password, HashFunc hashFunc)
+{
+    auto updateThreadProc = [this](const void * state)
```

Comment:
I think so - definitely something to look into. I think `PasswordHashUpdater` is instanced with `World` IIRC, so it could be destructed while work is queued. I might need to add something to purge the ThreadPool before `this` gets destructed.

Reply:
I ended up adding a shutdown method that gets called from `PasswordHashUpdater`. This allows for the threadpool to get shut down cleanly and prevent potential access violations if `this` gets destructed prior to the threadpool thread completing work.

Reply:
Nice!